src/Oxpecker/ModelBinder.fs

```diff
@@ -14,299 +12,259 @@ open Microsoft.Extensions.Primitives
+                | _ -> Seq.empty
+
+        match shapeof<'T> with
+        | Shape.String ->
+            fun _ (RawValueQuick values) -> values |> firstValue
+            |> wrap
+
+        | Shape.Nullable shape ->
+            shape.Accept
+                { new INullableVisitor<_> with
+                    member _.Visit<'t when 't: (new: unit -> 't) and 't: struct and 't :> ValueType>() = // 'T = Nullable<'t>
+                        let parse = mkParserCached<'t> ctx
+
+                        fun culture (RawValueQuick values) ->
+                            match values |> firstValue with
+                            | Null -> Nullable()
+                            | _ -> parse culture (StringValues.toDict values) |> Nullable
+                        |> wrap
+                }
+
+        | Shape.FSharpOption shape ->
+            shape.Element.Accept
+                { new ITypeVisitor<_> with
+                    member _.Visit<'t>() = // 'T = 't option
+                        let parse = mkParserCached<'t> ctx
+
+                        fun culture data ->
+                            match data with
+                            | RawValue values when values |> firstValue |> isNull -> None
+                            | _ -> parse culture data |> Some
+                        |> wrap
+                }
+
+        | Shape.FSharpList shape ->
+            shape.Element.Accept
+                { new ITypeVisitor<_> with
+                    member _.Visit<'t>() = // 'T = 't list
+                        let parse = mkParserCached<'t seq> ctx
+
+                        fun culture data -> parse culture data |> Seq.toList
+                        |> wrap
+                }
+
+        | Shape.Array shape when shape.Rank = 1 ->
+            shape.Element.Accept
+                { new ITypeVisitor<_> with
+                    member _.Visit<'t>() = // 'T = 't array
+                        let parse = mkParserCached<'t seq> ctx
+
+                        fun culture data -> parse culture data |> Seq.toArray
+                        |> wrap
+                }
+
+        | Shape.ResizeArray shape ->
+            shape.Element.Accept
+                { new ITypeVisitor<_> with
+                    member _.Visit<'t>() = // 'T = ResizeArray<'t>
+                        let parse = mkParserCached<'t seq> ctx
+
+                        fun culture data -> parse culture data |> ResizeArray
+                        |> wrap
+                }
+
+        | Shape.Enumerable shape ->
+            shape.Element.Accept
+                { new ITypeVisitor<_> with
+                    member _.Visit<'t>() = // 'T = 't seq
+                        if Type.(<>)(typeof<'T>, typeof<'t seq>) then
+                            unsupported typeof<'T>
+                        else
+                            mkParserCached<'t> ctx |> mkEnumerableParser |> wrap
+                }
+
+        | Shape.FSharpUnion(:? ShapeFSharpUnion<'T> as shape) ->
```

Comment:
Can Shape.NotStruct be added here similar to the Record shape matching? This should simplify visitor body

Reply:
Shape.NotStruct part was removed the record visitor. now, struct records can be parsed as well as struct unions. There is no reason to parse ref only unions and not struct unions